src/main/java/org/spongepowered/common/event/EventConsumer.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.event;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.spongepowered.api.event.Event;
+import org.spongepowered.common.SpongeImpl;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+public final class EventConsumer<E extends Event> {
+
+    public static <E extends Event> Builder<E> supplyEvent(Supplier<E> supplier) {
+        return new Builder<>(supplier.get());
+    }
+
+    public static <E extends Event> Builder<E> event(E event) {
+        return new Builder<>(event);
+    }
+
+    @Nullable private final Consumer<E> cancelledConsumer;
+    @Nullable private final Consumer<E> nonCancelledConsumer;
+    @Nullable private final Consumer<E> postConsumer;
+    private final E event;
+
+    private EventConsumer(Builder<E> builder) {
+        this.cancelledConsumer = builder.cancelledConsumer;
+        this.nonCancelledConsumer = builder.nonCancelledConsumer;
+        this.postConsumer = builder.postConsumer;
+        this.event = builder.event;
+    }
+
+    public static final class Builder<E extends Event> {
```

Comment:
It was just a quick write up, making things immutable and such. Overall, yes it could be cutting corners by directly having setters on the actual object.